winclip/winclip.1
.TH WINCLIP 1 "20 March 2016"
.\" (C) Copyright 1998-2016 Diomidis Spinellis.  All rights reserved.
.\"
.\" Permission to use, copy, and distribute this software and its
.\" documentation for any purpose and without fee is hereby granted,
.\" provided that the above copyright notice appear in all copies and that
.\" both that copyright notice and this permission notice appear in
.\" supporting documentation.
.\"
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.SH NAME
winclip \- console Windows clipboard interface
.SH SYNOPSIS
\fBwinclip\fP
[\fB\-h\fP|\fB\-v\fP]
[\fB\-w\fP|\fB\-u\fP|\fB\-m\fP|\fB\-g\fP|\fB\-r\fP]
[\fB\-b\fP]
[\fB\-l\fP \fIlang\fP]
[\fB\-s\fP \fIsublang\fP]
\fB-c\fP|\fB-p\fP|\fB-i\fP
[\fIfile\fP]
.SH DESCRIPTION
\fIWinclip\fP provides access to the Windows clipboard from a console
or MS-DOS window.
Its operation depends on its mandatory argument.
When executed with the -p (paste) argument it pastes
the Windows clipboard to its standard output or to the \fIfile\fP
optionally specified in the command line.
The \fIwinclip\fP output may be redirected
to a file (using the shell > syntax) or to another process
(using the shell | pipeline syntax).
When \fIwinclip\fP is executed with the -c (copy) argument
it copies its standard input (or the \fIfile\fP
optionally specified in the command line) to the Windows clipboard.
Again, \fIwinclip\fP input may be redirected to be read from a file
(using the < shell syntax), or from another process of a pipeline.
Finally, when \fIwinclip\fP is executed with the -i (information) argument
it will print on its standard output all data formats that can be retrieved
from the clipboard.
.LP
\fIWinclip\fP can currently paste text clipboard data, filenames
that have been copied in the Windows explorer, and bitmaps which
are pasted in binary \fIppm\fP format.
When tranfering text,
it will automatically perform the appropriate Windows to console
character code (OEM) set translations.
\fIWinclip\fP is an Intel architecture Win32 program that only works under
the Intel processor versions of Windows NT, 95, 98, Me, 2000, XP.
.SH OPTIONS
.IP "\fB\-c\fP"
Copy the standard input to the Windows clipboard.
.IP "\fB\-p\fP"
Copy the Windows clipboard contents to the standard output.
.IP "\fB\-i\fP"
Print on the standard output the data formats available in the clipboard.
Each format is displayed on a separate line.
The standard clipboard formats are printed using their Windows API CF_ names;
all other formats are printed using their registered name.
.IP "\fB\-w\fP"
Transfer text in the Windows-native format (CF_TEXT) instead of the
(default) DOS-native format (CF_OEMTEXT).
The Windows format is also known as ANSI text, but often represents
another Windows and locale-specific code page.
Setting this option will affect how non-ASCII characters will be encoded
for clipboard text and copied file icons.
.IP "\fB\-u\fP"
Transfer text in Unicode format (CF_UNICODETEXT) instead of the
(default) DOS-native format (CF_OEMTEXT).
The Unicode encoding is the same as the one used by the Windows clipboard:
apparently UCS-16 in little-endian byte order (UTF-16LE).
Setting this option will affect how non-ASCII characters will be encoded
for clipboard text and copied file icons.
.IP "\fB\-m\fP"
Transfer text in Unicode format (CF_UNICODETEXT) instead of the
(default) DOS-native format (CF_OEMTEXT).
The Unicode encoding is UTF-8.
Setting this option will affect how non-ASCII characters will be encoded
for clipboard text and copied file icons.
.IP "\fB\-r\fP"
Copy into the clipboard text specified to be in rich text format (RTF).
.IP "\fB\-b\fP"
When outputing Unicode text,
prepend a byte order mark (BOM) at the beginning.
This may make it easier for other tools to recognize \fIwinclip\fP's
output format.
When reading data for the clipboard, automatically detect
Unicode data, and set the encoding and endianess, if a byte-order-mark is
detected.
.IP "\fB\-l\fP \fBlang\fP"
Specify the language identifier for the text to be copied into
the clipboard.
The default value is \fINeutral\fP.
The identifier is case insensitive.
Specify the identifier \fIHelp\fP to obtain a list of supported
language settings.
.IP "\fB\-s\fP \fBsublang\fP"
Specify the sub-language identifier for the text to be copied into
the clipboard.
The default value is \fIDefault\fP.
The identifier is case insensitive.
Specify the identifier \fIHelp\fP to obtain a list of supported
sublanguage settings.
Specifying a sublanguage setting without the corresponding language setting
is probably not a good idea.
The combination of the default language and sublanguage identifiers in
effect result in the clipboard data getting marked as belonging to the
user default language.
.IP "\fB\-g\fP"
Specify that the data to be copied into the clipboard is a graphics image.
The program uses builtin code for loading
8-bit portable pixmap files (files starting with the P6 magic number),
and a Windows OLE interface for loading many other image files.
The OLE interface has been tested with JPEG, GIF, and BMP files.
.IP "\fB\-v\fP"
Print the program's version on the standard output and exit.
.IP "\fB\-h\fP"
Print a short help text on the program's operation on the standard output
and exit.
.SH EXAMPLE
dir /b | winclip -c
.br
Copies the directory list to the Windows clipboard.
.LP
winclip -p file.txt
.br
Saves the Windows clipboard to file.txt.
.LP
winclip -p | perl -pe "tr/[A-Z]/[a-z]/" | winclip -c
.br
Converts the contents of the Windows clipboard to lowercase.
.LP
winclip -c -l Portuguese -s Portuguese_Brazilian file.txt
.br
Copies the file file.txt to the Windows clipboard.
The clipboard's contents are marked to contain text in the language
Portuguese and sublanguage Portuguese_Brazilian.
.LP
pygmentize -f rtf code.js | winclip -cr
.br
Copies into the clipboard syntax-colored RTF generated by pygmentize.
.SH "SEE ALSO"
D. Spinellis.  Outwit: Unix tool-based programming meets the Windows world.
In \fIUSENIX 2000 Technical Conference Proceedings\fP, San Diego, CA, USA,
June 2000, USENIX Association.

.SH AUTHOR
(C) Copyright 1994-2005 Diomidis Spinellis.  All rights reserved.
.LP
Permission to use, copy, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation.
.LP
THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.SH BUGS
Should handle more clipboard content types.
.LP
\fIWinclip\fP may not work as expected if the code page of the console
window it is running is different from the code page of the clipboard
data or the code page it is specified to use.
You can check and change a console code page using the \fIchcp\fP command.
.LP
In operations involving 8-bit (non-Unicode) data,
the input locale (that's right, the active keyboard translation)
in effect at the time the text is placed in the clipboard
determines the character set associated with the clipboard data.
Therefore,
if you want to associate a Tamil or Greek code page with the
clipboard data, ensure that you have correctly set the active
keyboard type at the time you copy the data.


ChangeLog.txt
Version 1.25
(March 20 2016)

winclip supports the specification of copied data as RTF

winreg supports the REG_DWORD_BIG_ENDIAN data type

odbc now supports a -n option to specify the printed value of NULL fields.
(suggested by John McMahon)

winclip now supports a -g option for copying graphics files into the
clipboard.

The connection example in the odbc application is now correct
(suggested by Oliver Peters).

winclip now supports a -i option for reporting the available data formats.

Correct memory allocation and locking in winclip. (bug fix)

-----------------------------
Version 1.24
(February 10 2005)
New options in winclip allow the specification of the language
and sublanguage of the clipboard text.
-----------------------------
Version 1.23
(November 2004)
Made readlog more robust in the face of invalid data in the registry:
- Negative event time values would cause the program to crash.
- Error reporting on missing message sources was misleading.
-----------------------------
Version 1.22
(June 2004).
Change distribution format to zip files; distributed files reside an
a correctly-named directory.
-----------------------------
Version 21
(February 2004).
Update the winclip program to support
input and output of multi-byte (UTF-8) Unicode characters and
a byte order mark.

-----------------------------
Version 20
(December 2003)
Update the winreg program to support the
names with embedded backslashes.
Problem reported by Antti Ala-heikkil&auml;.

-----------------------------
Version 19
(December 2003)
Update the winreg program to support the
QWORD registry data type and add an option to ignore errors
and continue processing.

-----------------------------
Version 18
(December 2003)
Changes by Pascal Hirsch in the odbc program
corrected error reporting and allow direct connection to ODBC drivers
without requiring the registration of an ODBC data source.
The syntax for connecting to ODBC data sources has been slightly changed.

-----------------------------
Version 17
(November 2003)
A fix by Pascal Hirsch in the odbc program
fixes two access violations, and introduces better error reporting.

-----------------------------
Version 16
(September 2003)
Corrected erroneous pasting of non-Unicode text introduced
in version 15.

-----------------------------
Version 15
(September 2003)
Corrected problem of random characters appearing at the
end of winclip output.
Noted by Robert Bassett.

-----------------------------
Version 14
(June 2003)
Winreg can connect to remote machines over the network.
Thanks to Antti Ala-Heikkil&auml;.
-----------------------------
Version 13
(Jul 12 2002)
Added the readlog command;
winreg handles additional registry entry types.
-----------------------------
Version 12
(Jan 4 2002)
Winclip supports Unicode text;
gracefully handles non-ASCII filenames.
-----------------------------
Version 11
(Sep 11 2001)
Winclip accepts filename argument.
-----------------------------
Version 10
(Jun 1 2000)
First public release.
-----------------------------
Version 09
Review release.


winclip/winclip.c
/*
 * Copy/Paste the Windows Clipboard
 *
 * (C) Copyright 1994-2016 Diomidis Spinellis
 *
 * Permission to use, copy, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 */

#include <windows.h>
#include <malloc.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <io.h>
#include <wchar.h>

/* Add some formats lacking in VC5 */
#ifndef CF_DIBV5
#define CF_DIBV5 17
#endif

/* Size of an I/O chunk */
#define CHUNK 1024

/* Flags affecting Unicode output */
/* Print byte order mark */
static int bom = 0;
/* output UTF instead of native wide characters */
static int multibyte = 0;
/* True if big-endian BOM */
static int big_endian = 0;
/* True if graphics image bitmap */
static int bitmap = 0;

/* Other flags */
/* Text format */
static unsigned int textfmt = CF_OEMTEXT;
/* Language identifiers */
static int langid = LANG_NEUTRAL;
static int sublangid = SUBLANG_DEFAULT;

/* Program name */
char *argv0;

// The RTF clipboard format
static UINT cf_rtf;

static void
error(char *s)
{
	LPVOID lpMsgBuf;

	FormatMessage(
	    FORMAT_MESSAGE_ALLOCATE_BUFFER |
	    FORMAT_MESSAGE_FROM_SYSTEM |
	    FORMAT_MESSAGE_IGNORE_INSERTS,
	    NULL,
	    GetLastError(),
	    MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), /* Default language */
	    (LPTSTR) &lpMsgBuf,
	    0,
	    NULL
	);
	fprintf(stderr, "%s: %s: %s\n", argv0, s, lpMsgBuf);
	LocalFree(lpMsgBuf);
	exit(1);
}

void
unicode_fputs(const wchar_t *str, FILE *iofile)
{
	setmode(fileno(iofile), O_BINARY);

	if (multibyte) {
		int wlen = wcslen(str);
		int blen = (wlen + 1) * 6;
		char *mbs = malloc(blen);
		if (mbs == NULL) {
			CloseClipboard();
			fprintf(stderr, "%s: Unable to allocate %d bytes of memory\n", argv0, blen);
			exit(1);
		}
		if (WideCharToMultiByte(CP_UTF8, 0, str, wlen, mbs, blen, NULL, NULL) == 0) {
			CloseClipboard();
			error("Error converting wide characters into a multi byte sequence");
		}
		if (bom)
			fprintf(iofile, "\xef\xbb\xbf");
		fprintf(iofile, "%s", mbs);
	} else {
		if (bom)
			fprintf(iofile, "\xff\xfe");
		fwprintf(iofile, L"%s\n", str);
	}
}

static char *usage_string =
	"usage: %s [-v|h] [-w|u|m|g|r] [-l lang] [-s sublang] [-b] -c|-p|-i [filename]\n";

void
help(void)
{
	printf(usage_string, argv0);
	printf( "\t-v Display version and copyright information\n"
		"\t-h Display this help message\n"
		"\t-c Copy to clipboard\n"
		"\t-p Paste from clipboard\n"
		"\t-i Print the type of the clipboard's contents\n"
		"\t-u Data to be copied / pasted is in Unicode format\n"
		"\t-r Data to be copied in RTF format\n"
		"\t-m Unicode data is multi-byte\n"
		"\t-b Include BOM with Unicode data\n"
		"\t-w Data is in the Windows code page (OEM code page is the default)\n"
		"\t-l Specify the language for the data to be copied\n"
		"\t-s Specify the sublanguage for the data to be copied\n"
		"\t-g Data to be copied is a graphics bitmap image\n"
	);
	exit(0);
}

void
usage(void)
{
	fprintf(stderr, usage_string, argv0);
	fprintf(stderr, "Run %s -h for help\n", argv0);
	exit(1);
}

void
version(void)
{
	printf("%s - copy/paste the Windows clipboard.  $Revision: 1.25 $\n\n", argv0);
	printf( "(C) Copyright 1994-2016 Diomidis D. Spinelllis.  All rights reserved.\n\n"

		"Permission to use, copy, and distribute this software and its\n"
		"documentation for any purpose and without fee is hereby granted,\n"
		"provided that the above copyright notice appear in all copies and that\n"
		"both that copyright notice and this permission notice appear in\n"
		"supporting documentation.\n\n"

		"THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED\n"
		"WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF\n"
		"MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.\n\n"
	);
	exit(0);
}

struct s_keyval {
	char *name;
	int val;
};

#include "lang.h"

/*
 * Return a value for the passed key.
 * If the value is not found, print the valid keys and exit.
 */
int
get_val(const struct s_keyval *kv, int len, const char *key)
{
	int i, olen = 0;

	for (i = 0; i < len; i++)
		if (stricmp(kv[i].name, key) == 0)
			return (kv[i].val);
	fprintf(stderr, "%s: invalid key [%s].  Valid keys are:\n", argv0, key);
	for (i = 0; i < len; i++) {
		fprintf(stderr, "%s", kv[i].name);
		if ((olen += strlen(kv[i].name)) > 60) {
			fputc('\n', stderr);
			olen = 0;
		} else
			fputc(' ', stderr);
	}
	fputc('\n', stderr);
	exit(1);
}

/*
 * Return the name of a predefined clipboard format name.
 * Return NULL if this name is not known.
 *
 * This must be the intended implementation of the function
 * GetPredefinedClipboardFormatName, inexplicably appearing in
 * Microsoft's example.
 */
static const char *
predefined_name(int n)
{
	static char buff[256];

	if (n >= CF_GDIOBJFIRST && n <= CF_GDIOBJLAST) {
		sprintf(buff, "CF_GDIOBJ%d", n);
		return buff;
	} else if (n >= CF_PRIVATEFIRST && n <= CF_PRIVATELAST) {
		sprintf(buff, "CF_PRIVATE%d", n);
		return buff;
	}
	switch (n) {
	case CF_BITMAP:		 return "CF_BITMAP";
	case CF_DIB:		 return "CF_DIB";
	case CF_DIBV5:		 return "CF_DIBV5";
	case CF_DIF:		 return "CF_DIF";
	case CF_DSPBITMAP:	 return "CF_DSPBITMAP";
	case CF_DSPENHMETAFILE:	 return "CF_DSPENHMETAFILE";
	case CF_DSPMETAFILEPICT: return "CF_DSPMETAFILEPICT";
	case CF_DSPTEXT:	 return "CF_DSPTEXT";
	case CF_ENHMETAFILE:	 return "CF_ENHMETAFILE";
	case CF_HDROP:		 return "CF_HDROP";
	case CF_LOCALE:		 return "CF_LOCALE";
	case CF_METAFILEPICT:	 return "CF_METAFILEPICT";
	case CF_OEMTEXT:	 return "CF_OEMTEXT";
	case CF_OWNERDISPLAY:	 return "CF_OWNERDISPLAY";
	case CF_PALETTE:	 return "CF_PALETTE";
	case CF_PENDATA:	 return "CF_PENDATA";
	case CF_RIFF:		 return "CF_RIFF";
	case CF_SYLK:		 return "CF_SYLK";
	case CF_TEXT:		 return "CF_TEXT";
	case CF_WAVE:		 return "CF_WAVE";
	case CF_TIFF:		 return "CF_TIFF";
	case CF_UNICODETEXT:	 return "CF_UNICODETEXT";
	}
	return NULL;
}

/* Display the clipboard's formats */
static void
info(void)
{
	int n;

	if (!OpenClipboard(NULL))
		error("Unable to open clipboard");
	for (n = EnumClipboardFormats(0); n; n = EnumClipboardFormats(n)) {
		char buff[1024];
		LPCSTR fmtname;
		if ((fmtname = predefined_name(n)) == NULL) {
			if (GetClipboardFormatName(n, buff, sizeof(buff)) == 0)
				error("Unable to get clipboard format name");
			fmtname = buff;
		}
		printf("%s\n", fmtname);
	}
	if (GetLastError() != ERROR_SUCCESS)
		error("Error enumerating clipboard formats");
	CloseClipboard();
}

/*
 * Paste the Windows clipboard to standard output
 */
static void
paste(const char *fname, FILE *iofile)
{
	HANDLE cb_hnd;		/* Clipboard memory handle */

	if (!OpenClipboard(NULL))
		error("Unable to open clipboard");
	if (IsClipboardFormatAvailable(textfmt)) {
		/* Clipboard contains text; copy it */
		cb_hnd = GetClipboardData(textfmt);
		if (cb_hnd != NULL) {
			setmode(fileno(iofile), O_BINARY);
			if (textfmt == CF_UNICODETEXT)
				unicode_fputs(cb_hnd, iofile);
			else
				fprintf(iofile, "%s", cb_hnd);
		}
		CloseClipboard();
		exit(0);
	} else if (IsClipboardFormatAvailable(CF_HDROP)) {
		/* Clipboard contains file names; print them */
		cb_hnd = GetClipboardData(CF_HDROP);
		if (cb_hnd != NULL) {
			int nfiles, i;
			char fname[4096];

			nfiles = DragQueryFile(cb_hnd, 0xFFFFFFFF, NULL, 0);
			for (i = 0; i < nfiles; i++) {
				switch (textfmt) {
				case CF_TEXT:	/* Windows native */
					DragQueryFileA(cb_hnd, i, fname, sizeof(fname));
					fprintf(iofile, "%s\n", fname);
					break;
				case CF_UNICODETEXT:
					DragQueryFileW(cb_hnd, i, (LPWSTR)fname, sizeof(fname));
					unicode_fputs((wchar_t *)fname, iofile);
					break;
				case CF_OEMTEXT:
					DragQueryFileA(cb_hnd, i, fname, sizeof(fname));
					CharToOemBuff(fname, fname, strlen(fname));
					fprintf(iofile, "%s\n", fname);
					break;
				}
			}
		}
		CloseClipboard();
		exit(0);
	} else if (IsClipboardFormatAvailable(CF_BITMAP)) {
		/* Clipboard contains a bitmap; dump it */
		cb_hnd = GetClipboardData(CF_BITMAP);
		if (cb_hnd != NULL) {
			BITMAP bmp;
			HDC hdc;
			int x, y;
			COLORREF cref;
			HGDIOBJ oldobj;

			if (!GetObject(cb_hnd, sizeof(BITMAP), &bmp))
				error("Unable to get bitmap dimensions");
			setmode(fileno(iofile), O_BINARY );
			fprintf(iofile, "P6\n%d %d\n255\n", bmp.bmWidth, bmp.bmHeight);
			if ((hdc = CreateCompatibleDC(NULL)) == NULL)
				error("Unable to create a compatible device context");
			if ((oldobj = SelectObject(hdc, cb_hnd)) == NULL)
				error("Unable to select object");
			for (y = 0; y < bmp.bmHeight; y++)
				for (x = 0; x < bmp.bmWidth; x++) {
					cref = GetPixel(hdc, x, y);
					putc(GetRValue(cref), iofile);
					putc(GetGValue(cref), iofile);
					putc(GetBValue(cref), iofile);
				}
			SelectObject(hdc, oldobj);
			DeleteDC(hdc);
		}
		CloseClipboard();
	} else {
		CloseClipboard();
		error("The clipboard does not contain text or files");
	}
	if (ferror(iofile)) {
		perror(fname);
		exit(1);
	}
}

/*
 * Copy our input to the Windows Clipboard
 */
static void
copy_text(FILE *iofile)
{
	char *b, *bp;		/* Buffer, pointer to buffer insertion point */
	HANDLE cb_hnd;		/* Clipboard memory handle */
	HANDLE lc_hnd;		/* Locale memory handle */
	size_t bsiz, remsiz;	/* Buffer, size, remaining size */
	size_t total;		/* Total read */
	int n;
	char *clipdata;		/* Pointer to the clipboard data */
	LCID *plcid;		/* Locale id pointer */


	/* Read stdin into buffer */
	setmode(fileno(iofile), O_BINARY);
	bp = b = malloc(remsiz = bsiz = CHUNK);
	if (b == NULL) {
		CloseClipboard();
		fprintf(stderr, "%s: Unable to allocate %d bytes of memory\n", argv0, bsiz);
		exit(1);
	}
	total = 0;
	while ((n = fread(bp, 1, remsiz, iofile)) > 0) {
		remsiz -= n;
		total += n;
		if (remsiz < CHUNK) {
			b = realloc(b, bsiz *= 2);
			if (b == NULL) {
				CloseClipboard();
				fprintf(stderr, "%s: Unable to allocate %d bytes of memory\n", argv0, bsiz);
				exit(1);
			}
			remsiz = bsiz - total;
			bp = b + total;
		}
	}
	if (bom) {
		if (memcmp(b, "\xef\xbb\xbf", 3) == 0) {
			/* UTF-8 */
			textfmt = CF_UNICODETEXT;
			multibyte = 1;
			big_endian = 0;
		} else if (memcmp(b, "\xff\xfe", 2) == 0) {
			/* UTF-16 little endian (native) */
			textfmt = CF_UNICODETEXT;
			multibyte = 0;
			big_endian = 0;
		} else if (memcmp(b, "\xfe\xff", 2) == 0) {
			/* UTF-16 big endian */
			textfmt = CF_UNICODETEXT;
			multibyte = 0;
			big_endian = 1;
		}
	}
	if (textfmt == CF_UNICODETEXT && multibyte) {
		int b2siz = (total + 1) * 2;
		char *b2 = malloc(b2siz);
		int ret;

		if (b2 == NULL) {
			fprintf(stderr, "%s: Unable to allocate %d bytes of memory\n", argv0, b2siz);
			CloseClipboard();
			exit(1);
		}
		if ((ret = MultiByteToWideChar(CP_UTF8, 0, b, total, (LPWSTR)b2, b2siz)) == 0) {
			CloseClipboard();
			error("Error converting a multi byte sequence into wide characters");
		}
		total = ret * 2;
		b = b2;
	}
	/* Convert big to little endian */
	if (big_endian) {
		char tmp;
		unsigned i;

		for (i = 0; i < total; i++) {
			tmp = b[i];
			b[i] = b[i + 1];
			b[i + 1] = tmp;
		}
	}
	cb_hnd = GlobalAlloc(GMEM_MOVEABLE, total + 1);
	if (cb_hnd == NULL) {
		CloseClipboard();
		error("Unable to allocate clipboard memory");
	}
	clipdata = (char *)GlobalLock(cb_hnd);
	memcpy(clipdata, b, total);
	clipdata[total] = '\0';
	clipdata[total + 1] = '\0';
	GlobalUnlock(cb_hnd);
	SetClipboardData(textfmt, cb_hnd);
	/* Now set the locale */
	lc_hnd = GlobalAlloc(GHND, sizeof(LCID));
	if (lc_hnd == NULL) {
		CloseClipboard();
		error("Unable to allocate clipboard memory");
	}
	plcid = (LCID*)GlobalLock(lc_hnd);
	*plcid = MAKELCID(MAKELANGID(langid, sublangid), SORT_DEFAULT);
	GlobalUnlock(lc_hnd);
	SetClipboardData(CF_LOCALE, lc_hnd);
}


HBITMAP image_load(FILE * fp);

void
copy_bitmap(FILE *f)
{
	HBITMAP hb;
	hb = image_load(f);
	if (hb)
		SetClipboardData(CF_BITMAP, hb);
	else {
		CloseClipboard();
		error("Unable to load the graphics image bitmap");
	}
}

int getopt(int argc, char *argv[], char *optstring);
extern char	*optarg;	/* Global argument pointer. */
extern int	optind;		/* Global argv index. */

main(int argc, char *argv[])
{
	enum {O_NONE, O_COPY, O_PASTE, O_INFO} operation = O_NONE;
	int c;
	FILE *iofile;		/* File to use for I/O */
	char *fname;		/* and its file name */

	argv0 = argv[0];
	while ((c = getopt(argc, argv, "gvhuwcpimbl:rs:")) != EOF)
		switch (c) {
		case 'b':
			bom = 1;
			break;
		case 'm':
			if (textfmt != CF_OEMTEXT)
				usage();
			textfmt = CF_UNICODETEXT;
			multibyte = 1;
			break;
		case 'u':
			if (textfmt != CF_OEMTEXT)
				usage();
			textfmt = CF_UNICODETEXT;
			break;
		case 'r':
			if (textfmt != CF_OEMTEXT)
				usage();
			cf_rtf = RegisterClipboardFormat("Rich Text Format");
			if (cf_rtf == 0)
				error("Unable to register RTF clipboard format");
			textfmt = cf_rtf;
			break;
		case 'w':
			if (textfmt != CF_OEMTEXT)
				usage();
			textfmt = CF_TEXT;
			break;
		case 'c':
			if (operation != O_NONE)
				usage();
			operation = O_COPY;
			break;
		case 'p':
			if (operation != O_NONE)
				usage();
			operation = O_PASTE;
			break;
		case 'i':
			if (operation != O_NONE)
				usage();
			operation = O_INFO;
			break;
		case 'l':
			if (!optarg)
				usage();
			langid = get_val(lang, sizeof(lang) / sizeof(*lang), optarg);
			break;
		case 's':
			if (!optarg)
				usage();
			sublangid = get_val(sublang, sizeof(sublang) / sizeof(*sublang), optarg);
			break;
		case 'g':
			bitmap = TRUE;
			break;
		case 'v':
			version();
			break;
		case 'h':
			help();
			break;
		case '?':
			usage();
		}
	if (argc - optind > 1)
		usage();

	switch (operation) {
	case O_INFO:
		info();
		break;
	case O_PASTE:
		if (argc - optind == 1) {
			if ((iofile = fopen(fname = argv[optind], "w")) == NULL) {
				perror(argv[optind]);
				return (1);
			}
		} else {
			iofile = stdout;
			fname = "standard output";
		}
		paste(fname, iofile);
		break;
	case O_COPY:
		if (argc - optind == 1) {
			if ((iofile = fopen(fname = argv[optind], "r")) == NULL) {
				perror(argv[optind]);
				return (1);
			}
		} else {
			iofile = stdin;
			fname = "standard input";
		}
		if (!OpenClipboard(NULL))
			error("Unable to open the clipboard");
		if (!EmptyClipboard()) {
			CloseClipboard();
			error("Unable to empty the clipboard");
		}
		if (bitmap)
			copy_bitmap(iofile);
		else
			copy_text(iofile);
		CloseClipboard();
		if (ferror(iofile)) {
			perror(fname);
			exit(1);
		}
		break;
	case O_NONE:
		usage();
		break;
	}
	return (0);
}
